File: judgesystem/src/judgesystem/judges/experience.py
"""Experience requirement judge.

Evaluates whether a company's office has sufficient past project
experience to meet the bid announcement's experience requirements.
"""

from __future__ import annotations

import re
from datetime import datetime

import pandas as pd

from judgesystem.judges.base import BaseJudge
from judgesystem.models.judgement import (
    JudgementContext,
    JudgementResult,
    RequirementStatus,
)


class ExperienceJudge(BaseJudge):
    """Judge for experience requirements (実績要件)."""

    @property
    def requirement_type(self) -> str:
        return "実績要件"

    def evaluate(self, context: JudgementContext) -> JudgementResult:
        master = context.master_data
        office_no = context.office_no
        requirement_text = context.requirement_text

        achievements: pd.DataFrame = master.get("office_work_achievements", pd.DataFrame())
        construction: pd.DataFrame = master.get("construction", pd.DataFrame())
        agency: pd.DataFrame = master.get("agency", pd.DataFrame())

        # Filter achievements for this office
        office_achievements = achievements[achievements["office_no"] == office_no]

        if office_achievements.empty:
            return self._result(context, RequirementStatus.INSUFFICIENT,
                                "該当拠点の工事実績なし")

        # Extract conditions from requirement text
        conditions = self._extract_conditions(requirement_text)

        # Apply date filter if specified
        if conditions.get("year_from"):
            office_achievements = self._filter_by_date(
                office_achievements, conditions["year_from"]
            )
            if office_achievements.empty:
                return self._result(context, RequirementStatus.INSUFFICIENT,
                                    f"指定期間({conditions['year_from']}年以降)の実績なし")

        # Apply contractor layer filter
        if conditions.get("required_contractor_layer"):
            layer = conditions["required_contractor_layer"]
            office_achievements = office_achievements[
                office_achievements["contractor_layer"] == layer
            ]
            if office_achievements.empty:
                return self._result(context, RequirementStatus.INSUFFICIENT,
                                    f"指定立場({layer})の実績なし")

        # Apply construction type filter
        if conditions.get("construction_types"):
            office_achievements = self._filter_by_construction(
                office_achievements, conditions["construction_types"], construction
            )
            if office_achievements.empty:
                return self._result(context, RequirementStatus.INSUFFICIENT,
                                    "指定工種の実績なし")

        # Apply amount filter
        if conditions.get("min_amount"):
            office_achievements = office_achievements[
                office_achievements["total_amount"] >= conditions["min_amount"]
            ]
            if office_achievements.empty:
                return self._result(context, RequirementStatus.INSUFFICIENT,
                                    f"指定金額({conditions['min_amount']}円)以上の実績なし")

        return self._result(context, RequirementStatus.SUFFICIENT,
                            f"実績あり({len(office_achievements)}件)")

    def _extract_conditions(self, text: str) -> dict:
        """Extract structured conditions from requirement text."""
        conditions: dict = {}

        # Year extraction (e.g., "平成25年度以降" or "令和3年度以降")
        year_match = re.search(r"(平成|令和)(\d+)年度?以降", text)
        if year_match:
            era, year_num = year_match.group(1), int(year_match.group(2))
            if era == "平成":
                conditions["year_from"] = 1988 + year_num
            elif era == "令和":
                conditions["year_from"] = 2018 + year_num

        # Contractor layer
        if "元請" in text:
            conditions["required_contractor_layer"] = "元請け"

        # Amount
        amount_match = re.search(r"(\d[\d,]+)\s*(?:万円|千万円|億円|円)", text)
        if amount_match:
            raw = int(amount_match.group(1).replace(",", ""))
            if "億円" in text:
                raw *= 100_000_000
            elif "千万円" in text:
                raw *= 10_000_000
            elif "万円" in text:
                raw *= 10_000
            conditions["min_amount"] = raw

        return conditions

    def _filter_by_date(self, df: pd.DataFrame, year_from: int) -> pd.DataFrame:
        """Filter achievements by completion date >= year_from."""
        filtered = []
        for _, row in df.iterrows():
            try:
                completion = str(row.get("completion_date", ""))
                if completion and len(completion) >= 4:
                    year = int(completion[:4])
                    if year >= year_from:
                        filtered.append(row)
            except (ValueError, TypeError):
                continue
        if not filtered:
            return pd.DataFrame()
        return pd.DataFrame(filtered)

    def _filter_by_construction(
        self, df: pd.DataFrame, types: list[str], construction_master: pd.DataFrame
    ) -> pd.DataFrame:
        """Filter achievements by construction type."""
        if construction_master.empty:
            return df
        matching_nos = construction_master[
            construction_master["construction_name"].isin(types)
        ]["construction_no"].tolist()
        if not matching_nos:
            return df
        return df[df["construction_no"].isin(matching_nos)]

    def _result(
        self, ctx: JudgementContext, status: RequirementStatus, reason: str
    ) -> JudgementResult:
        return JudgementResult(
            announcement_no=ctx.announcement_no,
            company_no=ctx.company_no,
            office_no=ctx.office_no,
            requirement_type=self.requirement_type,
            status=status,
            reason=reason,
        )

File: lib/judgesystem/evaluators/experience.ts
/**
 * Experience Evaluator (実績要件)
 *
 * Evaluates whether a company's office has sufficient work experience
 * (completed projects) matching the announcement requirements.
 */

import type { EvaluationResult, OfficeWorkAchievement } from "@/lib/judgesystem/domain/types";
import type { Evaluator, EvaluatorContext } from "./base";

interface ExperienceConditions {
  yearFrom: number | null;
  requiredContractorLayer: string | null;
  requiresOriginalContractor: boolean;
  minAmount: number | null;
  minScore: number | null;
  requiredConstructionType: string | null;
  requiredAgencyNo: string | null;
}

function extractExperienceConditions(text: string): ExperienceConditions {
  const conditions: ExperienceConditions = {
    yearFrom: null,
    requiredContractorLayer: null,
    requiresOriginalContractor: false,
    minAmount: null,
    minScore: null,
    requiredConstructionType: null,
    requiredAgencyNo: null,
  };

  // Year condition: 令和X年度以降 or 平成X年度以降
  const reiwaMatch = text.match(/令和(\d+)年度以降/);
  if (reiwaMatch) {
    conditions.yearFrom = 2018 + parseInt(reiwaMatch[1], 10);
  }
  const heiseiMatch = text.match(/平成(\d+)年度以降/);
  if (heiseiMatch) {
    conditions.yearFrom = 1988 + parseInt(heiseiMatch[1], 10);
  }

  // Contractor layer
  if (/元請/.test(text)) {
    conditions.requiresOriginalContractor = true;
    conditions.requiredContractorLayer = "元請";
  } else if (/一次/.test(text)) {
    conditions.requiredContractorLayer = "一次下請";
  }

  // Minimum amount
  const amountMatch = text.match(/(\d[\d,]*)万円以上/);
  if (amountMatch) {
    conditions.minAmount = parseInt(amountMatch[1].replace(/,/g, ""), 10) * 10000;
  }

  // Minimum score
  const scoreMatch = text.match(/(\d+)点以上/);
  if (scoreMatch) {
    conditions.minScore = parseInt(scoreMatch[1], 10);
  }

  return conditions;
}

function matchesConditions(
  achievement: OfficeWorkAchievement,
  conditions: ExperienceConditions
): { matches: boolean; reason: string } {
  // Year check
  if (conditions.yearFrom && achievement.completionDate) {
    const completionYear = new Date(achievement.completionDate).getFullYear();
    if (completionYear < conditions.yearFrom) {
      return { matches: false, reason: `完了年度(${completionYear})が要件年度(${conditions.yearFrom})より前` };
    }
  }

  // Contractor layer check
  if (conditions.requiredContractorLayer) {
    if (achievement.contractorLayer !== conditions.requiredContractorLayer) {
      return { matches: false, reason: `請負階層(${achievement.contractorLayer})が要件(${conditions.requiredContractorLayer})と不一致` };
    }
  }

  // Amount check
  if (conditions.minAmount != null) {
    if (achievement.totalAmount < conditions.minAmount) {
      return { matches: false, reason: `契約金額(${achievement.totalAmount})が最低金額(${conditions.minAmount})未満` };
    }
  }

  // Score check
  if (conditions.minScore != null) {
    if (achievement.finalScore < conditions.minScore) {
      return { matches: false, reason: `工事成績(${achievement.finalScore})が最低点(${conditions.minScore})未満` };
    }
  }

  return { matches: true, reason: "条件充足" };
}

export class ExperienceEvaluator implements Evaluator {
  readonly type = "experience" as const;

  evaluate(ctx: EvaluatorContext): EvaluationResult {
    const { requirementText, officeNo, masterData } = ctx;

    const achievements = masterData.officeWorkAchievements.filter(
      (a) => a.officeNo === officeNo
    );

    if (achievements.length === 0) {
      return { isOk: false, reason: "実績要件：該当拠点の工事実績なし" };
    }

    const conditions = extractExperienceConditions(requirementText);

    const matchingAchievements = achievements.filter(
      (a) => matchesConditions(a, conditions).matches
    );

    if (matchingAchievements.length > 0) {
      return {
        isOk: true,
        reason: `実績要件：${matchingAchievements.length}件の適合実績あり`,
      };
    }

    return {
      isOk: false,
      reason: "実績要件：条件を満たす実績なし",
    };
  }
}
